OCLOMRS-383: add more sources and source search

ozezzy opened the pull request:
# JIRA TICKET NAME:
[add more selectable concept sources to addBulk concept page](https://issues.openmrs.org/browse/OCLOMRS-383)

# Summary:
user should be able to select a source other than CIEL when adding bulk concepts. there should be a search field to search for and select a source not listed as part of the radio buttons.

abulojoshua1 commented on src/components/bulkConcepts/addBulkConcepts.jsx:
```diff
@@ -35,24 +37,44 @@ export class AddBulkConcepts extends Component {
   componentDidMount() {
-    this.props.fetchCielConcepts();
+    this.props.getConceptSources();
+    this.props.fetchSourceConcepts(this.sourceUrl);
```
@ozezzy, kindly destructure props here.

ozezzy replied:
Destructuring getConceptSources and fetchSourcesConcepts here gives linting error because of line 11. They are already declared in the scope.

abulojoshua1 commented on src/components/bulkConcepts/addBulkConcepts.jsx:
```diff
@@ -35,24 +37,44 @@ export class AddBulkConcepts extends Component {
 
+  otherSourceClick = () => {
+    this.setState({ otherSelected: true });
+    this.props.getConceptSources();
```
@ozezzy, kindly destructure props here.

ozezzy replied:
destructuring  getConceptSources here give linting error

abulojoshua1 commented on src/components/bulkConcepts/addBulkConcepts.jsx:
```diff
@@ -35,24 +37,44 @@ export class AddBulkConcepts extends Component {
+
+  handleSourceSelect = (source) => {
+    this.sourceUrl = source.url.substring(1);
+    this.props.fetchSourceConcepts(this.sourceUrl);
```
@ozezzy, kindly destructure props here.

ozezzy replied:
Destructuring fetchSourceConcepts here will give a linting error

Tittoh commented:
@ozezzy Since we already have the CIEL radio button, do we need to have it in the dropdown too?

ozezzy commented:
> @ozezzy Since we already have the CIEL radio button, do we need to have it in the dropdown too?

CIEL is just added as a radio button for quick access, I don't see why to remove it from the list of available sources returned from the API (the dropdown). Let me know if removing it gives better usability though

dkayiwa commented on src/components/bulkConcepts/addBulkConcepts.jsx:
```diff
@@ -35,24 +37,44 @@ export class AddBulkConcepts extends Component {
     };
     this.invalidConceptIds = [];
-    this.sourceUrl = 'orgs/CIEL/sources/CIEL/concepts/';
+    this.sourceUrl = 'orgs/CIEL/sources/CIEL/';
```
What has this got to do with other sources other than CIEL?

ozezzy replied:
CIEL is the default when the page is loaded. `this.sourceUrl = 'orgs/CIEL/sources/CIEL/'` holds the sourceUrl, and it changes when a different source is selected.

dkayiwa commented on src/components/bulkConcepts/addBulkConcepts.jsx:
```diff
@@ -129,7 +152,7 @@ Dictionary
                 name="exampleRadios"
                 id="ciel"
-                value="option1"
+                value="/orgs/CIEL/sources/CIEL/"
```
What has this got to do with other sources other than CIEL?

ozezzy replied:
from my comment above, when a different source is selected, CIEL radio button is deselected, and `this.sourceUrl` changes to the url of the selected source. If CIEL radio button is selected agin, it gets the  the value `value="/orgs/CIEL/sources/CIEL/"` and assigns it to `this.sourceUrl` again.

dkayiwa commented:
The changes in this pull request seem to be doing more than a simple "add more selectable concept sources to addBulk concept page"

ozezzy commented:
It only added auto suggestion while searching on other sources, and modified the implementation of CIEL only format, to enable it work properly and reuse functions such as `validConcepts` lines 96 and 98 of `src/components/bulkConcepts/addBulkConcepts.jsx`

dkayiwa commented:
All changes in this pull request should simply be about ""add more selectable concept sources to addBulk concept page". Any other fixes or improvements should be another ticket and hence pull request.

ozezzy commented:
Ok. I will look through this, and see if there are changes not necessary in this PR

ozezzy commented:
I Think I better understand why it appears like there are unneeded changes. I made changes to initial state,  action and reducer functions that handled CIEL concept source to accommodate for other concept sources. I think this is justifiable in this case, as opposed to duplicating the set of action and reducer functions of the CIEL source implementation and changing the source URL.  Every change here is needed for this implementation to work. However, If this is not acceptable, I could close this PR and use the second approach. The resulting code would be much different from the current.

dkayiwa commented:
What would be the second approach?

ozezzy commented:
to create different sets of action and reducer functions, generally try to make minimal changes to the way it currently works for CIEL source. And then create a different ticket for improving the code

dkayiwa commented:
Which do you think is the better approach?

ozezzy commented:
I think the current approach is good in this case, It saves time, and the changes are all needed for the feature to work as implemented. Also considering the resulting code isn't too complex to reason along with.

dkayiwa commented:
In your video, you choose other and then select CIEL. Does this make sense when you already have CIEL up?

dkayiwa commented on src/components/bulkConcepts/addBulkConcepts.jsx:
```diff
@@ -138,6 +161,84 @@ Dictionary
+                <input
+                  className="form-check-input"
+                  type="radio"
+                  name="exampleRadios"
```
Did you intentionally call it exampleRadios?

ozezzy replied:
no, it is now fixed.

dkayiwa commented on src/components/bulkConcepts/addBulkConcepts.jsx:
```diff
@@ -138,6 +161,84 @@ Dictionary
+                  name="exampleRadios"
+                  id="otherSourcesOption"
+                  onClick={this.otherSourceClick}
+                  value="option3"
```
Can we get a better value other than option3?

ozezzy replied:
I have removed 'value'. it is not needed

dkayiwa commented on src/components/bulkConcepts/addBulkConcepts.jsx:
```diff
@@ -138,6 +161,84 @@ Dictionary
+                                    index,
+                                    item,
+                                    style: {
+                                      backgroundColor:
```
Can't we do this from styling file?

ozezzy replied:
No, it is passed as props to downshift directly, and contains logic

ozezzy commented:
> In your video, you choose other and then select CIEL. Does this make sense when you already have CIEL up?

I have fixed this and replaced the video

dkayiwa commented on src/components/bulkConcepts/addBulkConcepts.jsx:
```diff
@@ -8,17 +8,19 @@ import {
   Input,
 } from 'reactstrap';
-import fetchCielConcepts, { addExistingBulkConcepts, isConceptValid } from '../../redux/actions/bulkConcepts';
+import fetchSourceConcepts, { addExistingBulkConcepts, isConceptValid, getConceptSources } from '../../redux/actions/bulkConcepts';
```
Shouldn't the above be named fetchConceptSources

ozezzy replied:
fixed

dkayiwa commented on src/components/bulkConcepts/addBulkConcepts.jsx:
```diff
@@ -138,6 +161,84 @@ Dictionary
+                            ...getInputProps()
+                            }
+                            className="form-control search"
+                            id="sourceSearch"
```
Is the placeholder search with small s?

ozezzy replied:
fixed

dkayiwa commented on src/redux/actions/bulkConcepts/index.js:
```diff
@@ -1,21 +1,31 @@
+} from '../types';
 
-const fetchCielConcepts = () => async (dispatch) => {
+const fetchSourceConcepts = (url = 'orgs/CIEL/sources/CIEL/') => async (dispatch) => {
```
Are we hard coding this url value?

ozezzy replied:
fixed
